### PR TITLE
fix: internal dynamic categories using JSON

### PR DESCRIPTION
**Pull request by BeksOmega**
## The basics

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [X] I ran `npm run format` and `npm run lint`

### Resolves

#5580 

### Proposed Changes

Changes the internal dynamic categories to use JSON instead of XML.

Also moves registration of button callback to the workpace so they only get registered once, instead of every time the dynamic category opens.

Should be no change in behavior 🤞 

### Reason for Changes

XML system is iceboxed.

### Test Coverage

* Testing adding variables (including sort order) for both typed and untyped variable categories.
* Tested adding procedures.
* Tested adding parameters to procedures.

Tested on:
 * Desktop Chrome

### Documentation

N/A

### Additional Information

Existing category generators could not be removed because they are public. Deprecation warnings have been added to them.

### CODE REVIEW

**Comment by BeksOmega** [comment]
So I think the build is broken because `flyoutCategoryBlocks` is calling `generateVariableFieldDom` which is technically deprecated. But it needs to keep calling that for backwards compatibility, and I don't want to undeprecate it... so idk how to deal with this situation.

File: core/procedures.js [hunk 1]
```diff
@@ -294,6 +298,67 @@ const flyoutCategory = function(workspace) {
 };
 exports.flyoutCategory = flyoutCategory;
 
+/**
+ * Construct the blocks required by the flyout for the procedure category.
+ * @param {!WorkspaceSvg} workspace The workspace containing procedures.
+ * @return {!Array<!Object>} Array of objects defining the contents of the
+ *     category.
+ */
+const flyoutCategoryJson = function(workspace) {
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply core/procedures.js:307]: Actually, upon reading further I can see that `flyoutCategory` is used elsewhere too, so maybe it is best to be consistent.  I still think `Json` is probably not the best possible suffix, though.  🤔

File: core/procedures.js [hunk 2]
```diff
@@ -212,9 +213,12 @@ exports.rename = rename;
  * Construct the blocks required by the flyout for the procedure category.
  * @param {!WorkspaceSvg} workspace The workspace containing procedures.
  * @return {!Array<!Element>} Array of XML block elements.
+ * @deprecated
```
Comment (cpcallen) [inline core/procedures.js:216]: What should be  used instead?  (Here and every other place this tag appears.)

File: core/variables.js [hunk 3]
```diff
@@ -192,6 +199,123 @@ const flyoutCategoryBlocks = function(workspace) {
 };
 exports.flyoutCategoryBlocks = flyoutCategoryBlocks;
 
+/**
+ * Generate DOM objects representing a variable field.
+ * @param {!VariableModel} variableModel The variable model to
+ *     represent.
+ * @return {?Element} The generated DOM.
+ * @deprecated
```
Comment (cpcallen) [inline core/variables.js:207]: Here especially I think specifying what should be used instead might help you answer the question you pose in your PR comment.

File: core/variables.js [hunk 4]
```diff
@@ -192,6 +199,123 @@ const flyoutCategoryBlocks = function(workspace) {
...
+
+/**
+ * Construct the elements (blocks and button) required by the flyout for the
+ * variable category.
+ * @param {!WorkspaceSvg} workspace The workspace containing variables.
+ * @return {!Array<!Object>} Array of objects defining the contents of the
+ *     category.
+ * @package
+ */
+const flyoutCategoryJson = function(workspace) {
```
Comment (cpcallen) [inline core/variables.js:234]: See above.

File: core/workspace_svg.js [hunk 5]
```diff
@@ -488,21 +488,34 @@ class WorkspaceSvg extends Workspace {
     this.flyoutButtonCallbacks_ = Object.create(null);
 
     const Variables = goog.module.get('Blockly.Variables');
-    if (Variables && Variables.flyoutCategory) {
+    if (Variables) {
+      this.registerButtonCallback(
+          'CREATE_VARIABLE',
+          (button) => Variables.createVariableButtonHandler(
+              button.getTargetWorkspace()));
```
Comment (cpcallen) [inline core/workspace_svg.js:495]: Why was this call  to `.registerButtonCallback` not needed previously?  (Here and below.)

**Review by cpcallen — CHANGES_REQUESTED** [review]
Looks good except for some small details and the obvious issue of the build failure!

> So I think the build is broken because `flyoutCategoryBlocks` is calling `generateVariableFieldDom` which is technically deprecated. But it needs to keep calling that for backwards compatibility, and I don't want to undeprecate it... so idk how to deal with this situation.

Why is `generateVariableFieldDom`  deprecated, and what is the replacement?

* If it needs to be retained for backwards compatibility then it should probably not be deprecated.
* If it has a replacement, you should call that instead.
* If you really don't like either of those options and are feeling a bit chaotic neutral, you could just `@suppress` the diagnostic at the point that you make the call.

File: core/procedures.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (BeksOmega) [reply core/procedures.js:307]: I'm happy to change it to something else (I'm thinking of renaming the json version of `callbackFactory` in an upcoming PR). But if you want to retain it, I can't think of a better suffix than `Json` :/

File: core/workspace_svg.js (hunk 5 above)
Comment (BeksOmega) [reply core/workspace_svg.js:495]: It existed before, it was just called from `flyoutCategory` which would unnecessarily re-register the callback every time the flyout was opened. This is the proper place for the registrations.

**Comment by BeksOmega** [comment]
Ah ha ha! Wrt the build failure, turns out it shouldn't actually be a problem. The calls that are making the build fail are coming from the custom context menu options (which I'm fixing next) so I can just suppress the failures for now.

I thought they were coming from `flyoutCategoryBlocks` which I also just deprecated, and I was like... idk how to tell closure this is ok. Turns out it already knows! (as it should)

**Pull request closed without merging**